File: js/bin/transport_check.js
// Reads the shared transport-frame vector
// (spec/corpus/transcripts/transport-frame.json) and confirms this JS transport
// both reproduces the sealed ciphertext byte-for-byte and can open it again.
// The vector states both halves ("reproduces ct_hex and can open it"), so both
// are asserted: sealing alone would pass even if opening were broken, and
// opening alone would pass a transport that agreed with itself but not with the
// other implementations. Invoked by interop/check-transport-js.sh.
import fs from 'node:fs';
import { sealCiphertext, openCiphertext } from '../src/transport.js';

const path = process.argv[2];
if (!path) { process.stderr.write('usage: transport_check <transport-frame.json>\n'); process.exit(2); }
const { inputs, outputs } = JSON.parse(fs.readFileSync(path, 'utf8'));

let failures = 0;
const check = (name, want, got) => {
  if (got === want) { console.log(`PASS ${name}`); }
  else { console.log(`FAIL ${name}\n  got:  ${got}\n  want: ${want}`); failures++; }
};

const key = Buffer.from(inputs.key_hex, 'hex');
const seq = Number(inputs.seq);
const inner = Buffer.from(inputs.inner_plaintext_hex, 'hex');

check('ct_hex', outputs.ct_hex, sealCiphertext(key, seq, inner).toString('hex'));

const opened = openCiphertext(key, seq, Buffer.from(outputs.ct_hex, 'hex'));
if (opened === null) {
  console.log('FAIL inner_plaintext_hex\n  got:  <authentication failed>');
  failures++;
} else {
  check('inner_plaintext_hex', inputs.inner_plaintext_hex, opened.toString('hex'));
}

if (failures > 0) { process.stderr.write(`${failures} output(s) mismatched\n`); process.exit(1); }
console.log('transport frame seals and opens to the shared vector');

File: rust/src/bin/transport_check.rs
//! Cross-language interop check: reproduces the shared transport-frame vector
//! (spec/corpus/transcripts/transport-frame.json) with the Rust implementation,
//! and opens it again.
//!
//! The vector states both halves ("reproduces ct_hex and can open it"), so both
//! are asserted: sealing alone would pass even if opening were broken, and
//! opening alone would pass a transport that agreed with itself but not with the
//! other implementations. Invoked by interop/check-transport-rust.sh.

use std::process::exit;

use bonemesh::transport::{open_ciphertext, seal_ciphertext};
use serde_json::Value;

fn hex_decode(s: &str) -> Vec<u8> {
    (0..s.len()).step_by(2).map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap()).collect()
}

fn hex_encode(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

fn main() {
    let path = std::env::args().nth(1).expect("path to transport-frame.json");
    let doc: Value = serde_json::from_str(&std::fs::read_to_string(path).unwrap()).unwrap();
    let inputs = &doc["inputs"];
    let outputs = &doc["outputs"];

    let mut failures = 0;
    let mut check = |name: &str, want: &str, got: &str| {
        if want == got {
            println!("PASS {name}");
        } else {
            println!("FAIL {name}\n  got:  {got}\n  want: {want}");
            failures += 1;
        }
    };

    let key = hex_decode(inputs["key_hex"].as_str().unwrap());
    let seq = inputs["seq"].as_u64().unwrap();
    let inner_hex = inputs["inner_plaintext_hex"].as_str().unwrap();
    let inner = hex_decode(inner_hex);
    let want_ct = outputs["ct_hex"].as_str().unwrap();

    check("ct_hex", want_ct, &hex_encode(&seal_ciphertext(&key, seq, &inner)));

    match open_ciphertext(&key, seq, &hex_decode(want_ct)) {
        Some(pt) => check("inner_plaintext_hex", inner_hex, &hex_encode(&pt)),
        None => {
            println!("FAIL inner_plaintext_hex\n  got:  <authentication failed>");
            failures += 1;
        }
    }

    if failures > 0 {
        eprintln!("{failures} output(s) mismatched");
        exit(1);
    }
    println!("transport frame seals and opens to the shared vector");
}
